src/dune_rules/dune_file.ml

```diff
@@ -944,10 +944,88 @@ module Plugin = struct
 end
 
 module Install_conf = struct
+  module File_entry = struct
+    type t =
+      | Binding of File_binding.Unexpanded.t
+      | Include of
+          { context : Univ_map.t
+          ; path : String_with_vars.t
+          }
+
+    let decode =
+      let open Dune_lang.Decoder in
+      let decode_binding =
+        let+ binding = File_binding.Unexpanded.decode in
+        Binding binding
+      in
+      let decode_include =
+        sum
+          [ ( "include"
+            , let+ () = Syntax.since Stanza.syntax (3, 5)
+              and+ context = get_all
+              and+ path = String_with_vars.decode in
+              Include { context; path } )
+          ]
+      in
+      decode_binding <|> decode_include
+
+    let load_included_file path ~context =
+      let open Memo.O in
+      let+ contents =
+        Build_system.read_file (Path.build path) ~f:Io.read_file
+      in
+      let ast =
+        Dune_lang.Parser.parse_string contents ~mode:Single
+          ~fname:(Path.Build.to_string path)
+      in
+      let parse = Dune_lang.Decoder.parse decode context in
+      match ast with
+      | List (_loc, terms) -> List.map terms ~f:parse
+      | other ->
+        let loc = Dune_sexp.Ast.loc other in
+        User_error.raise ~loc [ Pp.textf "Expected list, got:\n%s" contents ]
+
+    let expand_include t ~expand_str ~dir =
+      let rec expand_include t ~seen =
+        match t with
+        | Binding binding -> Memo.return [ binding ]
+        | Include { context; path = path_sw } ->
+          let open Memo.O in
+          let* path =
+            expand_str path_sw
+            >>| Path.Build.relative
+                  ~error_loc:(String_with_vars.loc path_sw)
+                  dir
+          in
+          if Path.Build.Set.mem seen path then
+            User_error.raise
+              ~loc:(String_with_vars.loc path_sw)
+              [ Pp.textf "Include loop detected via: %s"
```

Comment:
We need to look into generalizing this cycle detection to have the same implementation for all includes.

Reply:
I added a module `Recursive_include` which adds `(include ...)` statements to config languages and used it to implement the include statement in the `(install ...)` stanza and also the `(include_dirs ...)` field.